xarray/core/dataset.py

```diff
@@ -101,6 +102,95 @@ def calculate_dimensions(variables):
     return dims
 
 
+def merge_indexes(indexers, variables, coord_names, append=False):
+    """Merge variables into multi-indexes.
+
+    Not public API. Used in Dataset and DataArray set_index
+    methods.
+    """
+    vars_to_replace = {}
+    vars_to_remove = []
+
+    for dim, var_names in indexers.items():
+        if isinstance(var_names, basestring):
+            var_names = [var_names]
+
+        names = []
+        arrays = []
+        current_index_variable = variables[dim]
+
+        if append:
+            current_index = current_index_variable.to_index()
+            if isinstance(current_index, pd.MultiIndex):
+                names.extend(current_index.names)
+                for i in range(current_index.nlevels):
+                    arrays.append(current_index.get_level_values(i))
```

Comment:
It might be premature optimization to worry about this, but something to watch out for is that this will refactorize the existing MultiIndex. It might be better to simply reuse existing `levels` and `labels`, if possible, and directly pass those into the `MultiIndex` constructor.

Note that to create `levels` and `labels` directly form an array (if necessary) you can use `Categorical` as used in `MultiIndex.from_arrays`, e.g.,
```python
cat = pd.Categorical(array, ordered=True)
levels = cat.categories
labels = cat.codes
```